### ISSUE TITLE
Error converting from i32 to f64 when type parameter implements From<f64>

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#73653]
I tried this code:

```rust
fn foo<T>(x: T, y: i32) -> f64 where f64: From<T> {
    f64::from(y)
}
```

I expected to see this happen: should compile and do the right thing.

Instead, this happened:

```rust
error[E0308]: mismatched types
 --> src/lib.rs:2:15
  |
1 | fn foo<T>(x: T, y: i32) -> f64 where f64: From<T> {
  |        - this type parameter
2 |     f64::from(y)
  |               ^ expected type parameter `T`, found `i32`
  |
  = note: expected type parameter `T`
                       found type `i32`
  = help: type parameters must be constrained to match other types
  = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
```

On the other hand, this code compiles fine:

```rust
fn foo<T>(_: T, y: i32) -> f64 {
    f64::from(y)
}
```

### Meta

I was surprised about this, so after running this with 1.44.0 and 1.44.1 I checked older versions, but 1.39.0, 1.31.0 and 1.15.0 all give the same error. Seems likely that there is an issue about this already, but no clue how to look for it.

**Comment by lzutao** [comment]
I don't know if we can bound on primitive types like `f64`. But if you bound `T` by `Into` trait, this should compile:
```rust
fn foo<T>(x: T, y: i32) -> f64 where T: Into<f64>;
```

**Comment by lcnr** [comment]
Might be related to https://github.com/rust-lang/rust/issues/34979#issuecomment-619505332

I think this is caused by winnowing somehow incorrectly discarding a relevant predicate.

cc @eddyb @nikomatsakis for good measure

**Comment by nikomatsakis** [comment]
Yeah, I agree with @lcnr's diagnosis. This is an example of a known limitation with the current trait solver, sort of a sub-issue of https://github.com/rust-lang/rust/issues/41756 or some other similar issues. I'm not sure that we have any plans to fix this apart from transitioning over to chalk.
